### PR TITLE
Improve error checking for OIDC/SAML mapping providers

### PR DESCRIPTION
**Pull request by clokep**
~~Additional tests for the feature added in #8345. The tests are a bit funky since the case-sensitive logic only really works for lowercase localparts (this also seems to be true of the similar code in the SAML module) due to the use of `map_username_to_mxid_localpart` which automatically lower-cases things.~~

This does a couple of unrelated things:

* Extends the OIDC tests for existing users and invalid data.
* Checks that the localpart returned by mapping providers for SAML and OIDC are valid before registering new users.
* ~~Adds some very basic SAML tests.~~ I'll move these to a separate PR.

### CODE REVIEW

File: tests/handlers/test_oidc.py [hunk 1]
```diff
@@ -705,18 +705,62 @@ def test_map_userinfo_to_user(self):
     def test_map_userinfo_to_existing_user(self):
         """Existing users can log in with OpenID Connect when allow_existing_users is True."""
         store = self.hs.get_datastore()
-        user4 = UserID.from_string("@test_user_4:test")
+        user = UserID.from_string("@test_user:test")
```
Comment (clokep) [inline tests/handlers/test_oidc.py:708]: There was no reason to start a `4` here so I simplified.

File: tests/handlers/test_oidc.py [hunk 2]
```diff
@@ -705,18 +705,65 @@ def test_map_userinfo_to_user(self):
...
+
+        # Logging in when matching a name exactly should work.
+        #
+        # Note that this essentially means matching an existing all lowercase
+        # name, regardless of the case of the localpart providing by the mapping
+        # provider.
+        #
+        # Above we return test_USER_2, which you'd expect to match the user
+        # test_USER_2, which was already created, but it actually ends up
+        # matching test_user_2.
```
Comment (clokep) [inline tests/handlers/test_oidc.py:758]: This really makes me think the code is not working as expected. In the case of SAML it is left up to the mapping provider to generate a "valid" MXID. For SAML, the `map_username_to_mxid_localpart` is only used for "grandfathering old data" and the default mapping provider uses dot replace or hex-encode (see #6411).

**Comment by clokep** [comment]
@anoadragon453 and I talked a bit about this in #synapse-dev:matrix.org and it seems the current behavior isn't really what is wanted. I'll try to re-work this so the case-sensitivity match works properly.

File: tests/handlers/test_oidc.py (hunk 2 above)
Comment (clokep) [reply tests/handlers/test_oidc.py]: This was made simpler by pushing the normalization into the mapper.

File: synapse/handlers/oidc_handler.py [hunk 3]
```diff
@@ -885,10 +890,12 @@ async def _map_userinfo_to_user(
             "Retrieved user attributes from user mapping provider: %r", attributes
         )
 
-        if not attributes["localpart"]:
-            raise MappingException("localpart is empty")
-
-        localpart = map_username_to_mxid_localpart(attributes["localpart"])
```
Comment (clokep) [inline synapse/handlers/oidc_handler.py:891]: This got pushed into the mapping provider, which matches SAML and gives mapping providers a bit more control (e.g. the dot-replace vs. hex-encode options that SAML has).

**Comment by clokep** [comment]
I've removed the SAML tests for now since I don't want to block this on messing with our buildkit envs. I'll put them up as a separate PR.

File: synapse/handlers/oidc_handler.py (hunk 3 above)
Comment (erikjohnston) [reply synapse/handlers/oidc_handler.py:891]: Is that a change in the mapping provider API?

Comment (clokep) [reply synapse/handlers/oidc_handler.py:891]: The SAML mapping provider (implicitly) must provide a valid localpart. I think this was kind of the assumption with the OIDC one as well, but I'm not sure. I failed to update the documentation to make this explicit for both of them, but I meant to! 😄

Comment (clokep) [reply synapse/handlers/oidc_handler.py:891]: I pushed some documentation improvements (and added something to `UPGRADE.rst`). I suspect not many people have their own mapping provider, but I'm not sure.

**Review by erikjohnston — APPROVED** [review]

**Pull request merged**